### PR TITLE
[NVIDIA XLA GPU] Turn the cudnn fused attention feature ON by default

### PR DESCRIPTION
**Pull request by AyanmoI**
Now that CUDA 12 and cuDNN 8.9.4 are now in CI and the feature is in a steady state, it is time to turn this ON by default and reap the benefits.
Note that this doesn't include Flash Attention.

### CODE REVIEW

**Comment by AyanmoI** [comment]
We would like to keep an eye on the CI to make sure that the fused attention pipeline has no side-effects.

**Comment by thomasjoerg** [comment]
Thank you for the PR, Ayan. Let me benchmark this change enabled and get back to you.

**Review by thomasjoerg — APPROVED** [review]

**Comment by AyanmoI** [comment]
Any other blockers to this to get merged?

**Comment by thomasjoerg** [comment]
Hi Ayan, I see larger wins on a few benchmarks but also small regressions across many others. The geomean comes out slightly negative for the overall benchmark suite. We need a bit more time to generate reproducers that we can share. how much urgency is there on your end?

**Comment by AyanmoI** [comment]
> Hi Ayan, I see larger wins on a few benchmarks but also small regressions across many others. The geomean comes out slightly negative for the overall benchmark suite. We need a bit more time to generate reproducers that we can share. how much urgency is there on your end?

Thanks @thomasjoerg :).Let's proceed with a thorough investigation. I was just wondering if there were any updates on the benchmarks. Please keep me posted.

**Comment by AyanmoI** [comment]
@thomasjoerg Any updates on this?

**Comment by thomasjoerg** [comment]
Sorry to keep you waiting. I discovered an issue, which was unrelated to this PR and am re-running benchmarks atm.

**Comment by thomasjoerg** [comment]
First batch of benchmark are looking good now. There's a few more I want to take a look at. I'll report back on Wednesday (will be out tomorrow).

**Comment by AyanmoI** [comment]
> First batch of benchmark are looking good now. 
That is good to know :)
> There's a few more I want to take a look at. I'll report back on Wednesday (will be out tomorrow).
Sounds good!

**Comment by thomasjoerg** [comment]
Update: `set_xla_gpu_enable_cudnn_fmha(true)` seems to cause Check failure in some models. Investigating ...

**Comment by thomasjoerg** [comment]
Quick update: I verified that `--xla_gpu_enable_cudnn_fmha=true` crashes the compiler on a legitimate input HLO module. Unfortunately, I cannot share the full unoptimized HLO of this Google-internal model. I am narrowing down the issue and will get back with details / a minimal reproducer as soon as I can.

**Comment by AyanmoI** [comment]
> Quick update: I verified that `--xla_gpu_enable_cudnn_fmha=true` crashes the compiler on a legitimate input HLO module. Unfortunately, I cannot share the full unoptimized HLO of this Google-internal model. I am narrowing down the issue and will get back with details / a minimal reproducer as soon as I can.

Hey Thomas, any updates on the reproducer?

**Comment by thomasjoerg** [comment]
Debugging this was quite a rabbit hole. The `CudnnFusedMHARewriter` seems to introduce `transpose` ops that trigger an otherwise unrelated bug in GEMM rewrite / autotuning passes that run at a later point in time. I made a reproducer and we are working on a fix. Sorry this is taking so long.

**Comment by AyanmoI** [comment]
> Debugging this was quite a rabbit hole. The `CudnnFusedMHARewriter` seems to introduce `transpose` ops that trigger an otherwise unrelated bug in GEMM rewrite / autotuning passes that run at a later point in time. I made a reproducer and we are working on a fix. Sorry this is taking so long.

Ah! That's a bummer. We could have helped you debug this. Let me know if I can assist in anyway.

**Comment by thomasjoerg** [comment]
> Ah! That's a bummer. We could have helped you debug this. Let me know if I can assist in anyway.

Unfortunately, our current open-source development story makes this hard sometimes. This was one of those cases. The reason is that Google internally builds from head. Therefore all changes need significant pre-merge testing and benchmarking. When breakages or regressions arise in Google-internal models, we may be unable to share details. We try to make minimal reproducers and have some tools to help, however, in this case making a reproducer was almost all the work. Once we had it, the fix was quick. 

@puneith is monitoring the situation and interested in making community contributions much easier. I realize we have some way to go here. In my defense, while this PR is small it enables PR #60233 which added 5k LOC. Changes of this size are tricky to merge, because the blast radius is inherently high. My team and I try to provide the support to the community to make that feasible as good as we can.

**Comment by thomasjoerg** [comment]
Things are looking promising. I'm hopeful we can merge the PR tomorrow or the day after. I'll be in touch.

**Comment by puneith** [comment]
> Things are looking promising. I'm hopeful we can merge the PR tomorrow or the day after. I'll be in touch.

I think PRs like these might be a good use case for the conversation we're already having with NVIDIA on PR benchmarking regression analysis. As @thomasjoerg mentions we have some way to go (particularly given our internal benchmarks), but I think this is a good opportunity to initiate work on solving for these cases.

**Comment by xla-rotation** [comment]
This is a hard crash IIUC - so I don't think it's a good example where a
separate default flag would've helped.

**Comment by AyanmoI** [comment]
> Unfortunately, our current open-source development story makes this hard sometimes. This was one of those cases. The reason is that Google internally builds from head. Therefore all changes need significant pre-merge testing and benchmarking. When breakages or regressions arise in Google-internal models, we may be unable to share details. 
That is understandable and it's great that we are moving towards, slowly but surely, improving this development and validation workflow in OSS.
>We try to make minimal reproducers and have some tools to help, however, in this case making a reproducer was almost all the work. Once we had it, the fix was quick.
Thanks for taking care of this though :)

> @puneith is monitoring the situation and interested in making community contributions much easier. I realize we have some way to go here. 
Yes! As @puneith mentioned, this is a good use case.

> In my defense, while this PR is small it enables PR #60233 which added 5k LOC. Changes of this size are tricky to merge, because the blast radius is inherently high. My team and I try to provide the support to the community to make that feasible as good as we can.
Well, in fact its larger than 5k LOC including all the PRs :)

**Comment by thomasjoerg** [comment]
Presubmit checks run into 3 test breakages. 

[TritonGemmTest.SlicedBatchDimensionIsSupported](https://github.com/openxla/xla/blob/db126aaef97a54a1a6cb7a86fa9c21cb96b5b2dc/xla/service/gpu/ir_emitter_triton_test.cc#L1340) fails.

[int4_test.cc - NonMajorToMinorLayout](https://github.com/openxla/xla/blob/db126aaef97a54a1a6cb7a86fa9c21cb96b5b2dc/xla/tests/int4_test.cc#L64)

Please let me know if have issues to reproduce.

There is another test breakage in the TensorFlow repo. The following numerical assertion fails with `AssertionError: 0.19496745 not less than 0.01`

https://github.com/tensorflow/tensorflow/blob/864420736135828e6326dfc5e016b46720996088/tensorflow/compiler/tests/svd_op_test.py#L59

Strictly speaking, you don't have to worry about downstream tests in TensorFlow, but given that it looks like a numerical issue, your help would be much appreciated!

**Comment by AyanmoI** [comment]
Thanks @thomasjoerg . I was OOO for the past week. I'll start looking at this asap.

**Comment by thomasjoerg** [comment]
Thank you, Ayan!

Please let me know if you can reproduce the test breakages. The CI on this PR seems fine, but I believe the tests have been added / touched recently. Consider to sync and kick off another CI run first. I can help with the latter if needed.

**Comment by AyanmoI** [comment]
@thomasjoerg I was able to reproduce the failures locally. However, in the rerun of CI, I only see the `[int4_test.cc](https://github.com/openxla/xla/blob/db126aaef97a54a1a6cb7a86fa9c21cb96b5b2dc/xla/tests/int4_test.cc#L64)` failure but not the TritonGemmTest. Not sure why. `//xla/service/gpu:ir_emitter_triton_test_gpu                             PASSED in 21.2s` seems to have passed.

**Comment by AyanmoI** [comment]
@thomasjoerg Also, you mentioned there being some internal failures. I assume those are the ones you have already fixed. These OSS test failures are the only one that remain. Is that correct?

**Comment by thomasjoerg** [comment]
> These OSS test failures are the only one that remain. Is that correct?

Yes, these were the only failures that spotted in pre-submit checks. We took care of anything else already.

**Comment by radhakrishnaba** [comment]
Hi @AyanmoI, can you have a look at the comments? Thanks.

**Comment by AyanmoI** [comment]
> Hi @AyanmoI, can you have a look at the comments? Thanks.

Yea. Already on it.

**Comment by AyanmoI** [comment]
@thomasjoerg Fixed the issues. Sorry couldn't get to it sooner. Hopefully this does it.

**Comment by akuegel** [comment]
So I see you committed a fix for the ir_emitter_triton_test. What about the int4_test, and the test with numerical differences in Tensorflow?

**Comment by AyanmoI** [comment]
> So I see you committed a fix for the ir_emitter_triton_test. What about the int4_test, and the test with numerical differences in Tensorflow?

I didn't need to do anything to fix the int4_test since the issue was mainly the S4/U4 verifier that has been removed. I haven't been able to look into the TF test. Does that still fail? I can't see that in the CI.

**Comment by akuegel** [comment]
> I didn't need to do anything to fix the int4_test since the issue was mainly the S4/U4 verifier that has been removed. I haven't been able to look into the TF test. Does that still fail? I can't see that in the CI.

Thanks, I confirmed the int4_test is passing now.
Yes, the svd_ops_test still fails. I debugged it a bit, and it is related to running LayoutNormalization. But it works if (before running LayoutNormalization) we run the MoveCopyToUsers pass. My guess is that the test tolerance is too strict, and so we get different numeric results based on slightly different HLO.

**Review by akuegel — APPROVED** [review]

File: xla/service/gpu/ir_emitter_triton_test.cc
Comment (akuegel) [inline xla/service/gpu/ir_emitter_triton_test.cc:1571]: It turns out that this will not be necessary when we add the MoveCopyToUsers pass before running LayoutNormalization.

**Comment by akuegel** [comment]
Unfortunately this will have to be rolled back. An internal test revealed a bug:

https://github.com/openxla/xla/blob/7063061ecb2ae315623a9d9ee2e907e17c5a07a0/xla/service/gpu/cudnn_fused_mha_rewriter.cc#L1141

Here you change the dimension numbers of a dot without changing its operands. This results in invalid HLO, and unless you make sure that dot will be removed, things will fail. In the reproducer, the dot stays and then we run into a check failure later in the pipeline.

I have tried a kind of hacky fix here: https://github.com/openxla/xla/pull/6958
This would avoid creating invalid HLO, but it will make 4 tests fail. But I think it shows the general idea what you need to do to make it work correctly.

I have asked whether I am allowed to share the HLO reproducer with you.

**Comment by akuegel** [comment]
@AyanmoI I have minimized the reproducer now, still waiting for the approval to share. But the general idea is that it is HLO that is matched to custom_call_target="__cudnn$fhmaScaleBiasSoftmax", and the "first" dot will not be removed by DCE because the divide op of the softmax has another user (but I believe we would see the same issue if it is one of the other ops on the path from the first dot to the second dot that has another user). The question is whether matching such HLO makes sense. It essentially means we would duplicate the HLO and run it once with custom call, and another time without the custom call (but also without the final dot).

**Pull request closed without merging**
